app/src/main/java/com/github/multimatum_team/multimatum/repository/FirebaseGroupRepository.kt

```diff
@@ -0,0 +1,108 @@
+package com.github.multimatum_team.multimatum.repository
+
+import android.util.Log
+import com.github.multimatum_team.multimatum.model.GroupID
+import com.github.multimatum_team.multimatum.model.UserGroup
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.ktx.firestore
+import com.google.firebase.ktx.Firebase
+import kotlinx.coroutines.tasks.await
+
+/**
+ * Remote Firebase repository for storing user groups.
+ */
+class FirebaseGroupRepository : GroupRepository() {
+    private var database: FirebaseFirestore = Firebase.firestore
```

Comment:
Couldn't `database` be a `val`?

Reply:
Also, you say that it is almost impossible to test some parts of the code. Wouldn't it be possible to inject `database` in this class, so that we could generate a mock database using Mockito? If it works it would fix a large part of the problem as this class has 0% coverage.

Reply:
Possibly, it could increase the coverage but wouldn't actually test anything meaningful. Another more concerning issue is that, from what I understand, mockito only lets you specify what is returned from certain method calls of the mocked object, e.g. `when(mockedObject.method(arg1, arg2)).thenReturn(result)` but in our case we have to mock chains of method calls: consider the `fetch` implementation in `FirebaseGroupRepository` (which is the simplest method in the class):
```kt
    override suspend fun fetch(id: GroupID): UserGroup =
        deserializeGroup(database.collection("groups").document(id).get().await())
```
If we were to mock `FirebaseFirestore`, then wouldn't we have to do something like this?
```ks
val mockSnapshot = mock(DocumentSnapshot::class.java)
when(mockSnapshot.get("name")).thenReturn("test name")
... // same for owner and members
val mockDocument = mock(DocumentReference::class.java)
when(mockDocument.get()).thenReturn(mockSnapshot)
val mockCollection = mock(CollectionReference::class.java)
when(mockCollection.document(dummyID)).thenReturn(mockDocument)
val mockDatabase = mock(FirebaseFirestore::class.java)
when(mockDatabase.collection("groups")).thenReturn(mockCollection)
```

Reply:
Yes I think that you would have to do that. I am not sure of how many such "chains" of mock you would have to do, maybe it is feasible (assuming one person is tasked with this and if needed spends a whole sprint on it)? Because achieving 80% coverage is not easy and we cannot afford that a whole large class is not tested at all... Otherwise you may be able to do some (ugly, but meaningful) tests using reflection?

Reply:
I wrote a draft and it indeed seems to be working, but is extremely tedious. Perhaps we can automate the process by mocking the entire firebase database from a hashmap?

Reply:
This would probably make a lot of sens, yes

Reply:
I got a 11% increase with my dumb test, I think I can reach 50% without too much trouble.

Reply:
I wrote some helper functions to generate the mock firebase database as discussed above, and managed to reach 44% diff-coverage by testing `fetchAll`. Unfortunately, I'm not sure I see how to go further and mock `add` and `delete`, which require changing the state of the database. I don't see a way to mock `addSnapshotListener` either.
I'll try to investigate how to implement this tomorrow, but if I don't succeed, maybe we can still remove the 50% threshold temporarily, and I'll make up for the -1.5% total coverage by applying the same technique for FirebaseDeadlineRepository.

Reply:
I don't exactly get the problem. What `add` and `delete` do you need to mock?

Reply:
> I don't exactly get the problem. What `add` and `delete` do you need to mock?

The ones from Firebase, specifically `CollectionReference.add` and `DocumentReference.delete`